perf: refactor imbuement decay to per-item management

dudantas opened the pull request:
# Description

This refactors the imbuement decay system, changing it from a player-wide loop to a per-item approach. This significantly optimizes performance by reducing redundant checks and focusing only on items that require decay updates. The refactor includes:
- Dynamic scheduling based on active imbuements.
- Handling decay conditions for protection zones, fight mode, and backpack rules directly on individual items.
- Improved scalability and maintainability of the code.

Resolves #3135

## Behaviour
### **Actual**
The previous implementation iterated through all players and their inventory items to process imbuement decay. This led to unnecessary overhead, even for players with no relevant items.

### **Expected**
The new implementation tracks only items with active imbuements, dynamically managing their decay schedule and conditions. This minimizes unnecessary processing and ensures accurate decay handling.

## Type of change
- [x] Performance improvement (non-breaking change that improves efficiency)
- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested

The changes were tested with the following scenarios:
- **Test A:** Items with imbuements were observed for accurate decay, ensuring decay stops or continues correctly based on conditions.
- **Test B:** Players in protection zones and fight mode were tested to ensure appropriate imbuement handling.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I checked the PR checks reports
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Centralized imbuement decay subsystem with per-item tracking; decay now starts when imbuements are applied and on player login.

* **Bug Fixes**
  * Decay is tied to equip/dequip and per-item lifecycle, removing the previous global periodic inventory refresh and improving accuracy.
  * Minor event identifier adjustment for scheduled tasks.

* **Tests**
  * Added unit/integration tests and test fixtures covering imbuement data and decay behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dudantas committed:
perf: refactor imbuement decay to per-item management

Refactored the imbuement decay system to manage individual items directly instead of iterating through all players. This improves performance by:

Tracking only items with active imbuements.
Scheduling events dynamically based on items needing updates.
Reducing redundant checks and player-wide loops.
Simplifying imbuement decay logic while preserving existing functionality, including protection zone, fight mode, and backpack-specific rules.
This change optimizes the handling of imbuements and ensures better scalability.

dudantas committed:
fix: correct ++ slot id

lBaah commented:
Function `stopImbuementDecay` need to check if it can decay imbuement duration before doing it, otherwise it will decay the imbuement time when you unequip an item from the inventory, no matter if it can or not decay, based on the elapsed time. For example, it decays when in protection zones.

dudantas commented:
> Function `stopImbuementDecay` need to check if it can decay imbuement duration before doing it, otherwise it will decay the imbuement time when you unequip an item from the inventory, no matter if it can or not decay, based on the elapsed time. For example, it decays when in protection zones.

I didn't understand your comment. Have you tested it? This check has already been done.

FelipePaluco committed:
Merge branch 'main' into dudantas/perf-check-imbuements

gabrielew commented:
> > Function `stopImbuementDecay` need to check if it can decay imbuement duration before doing it, otherwise it will decay the imbuement time when you unequip an item from the inventory, no matter if it can or not decay, based on the elapsed time. For example, it decays when in protection zones.
> 
> I didn't understand your comment. Have you tested it? This check has already been done.
@dudantas 

- unequip/equip decay 1 minute of imbuement time
- minutes is flickering
![client_-_local_lnZCeRQiXy](https://github.com/user-attachments/assets/eeb8e35c-4f6d-423d-9f04-7bb3bfb24bdc)


The imbuement time doesn't decay in combat
![image](https://github.com/user-attachments/assets/5d8e9b8d-858e-4bef-babb-d67ec9fbacb6)

edit: is decaying without battle, but when we enter in protected zone the time back to the start

**log on player login**
```txt
[2025-01-07 10:38:02.165] [info] Starting imbuement decay for item balrog backpack
[2025-01-07 10:38:02.165] [info] Starting imbuement decay for item balrog backpack
[2025-01-07 10:38:02.165] [info] Starting imbuement decay for item balrog backpack
[2025-01-07 10:38:02.166] [info] Starting imbuement decay for item balrog backpack
[2025-01-07 10:38:02.166] [info] Starting imbuement decay for item balrog backpack
[2025-01-07 10:38:02.193] [info] Starting imbuement decay for item alicorn headguard
[2025-01-07 10:38:02.193] [info] Starting imbuement decay for item alicorn headguard
[2025-01-07 10:38:02.193] [info] Starting imbuement decay for item balrog backpack
[2025-01-07 10:38:02.194] [info] Starting imbuement decay for item soulshell
[2025-01-07 10:38:02.194] [info] Starting imbuement decay for item soulshell
[2025-01-07 10:38:02.194] [info] Starting imbuement decay for item soulbleeder
[2025-01-07 10:38:02.195] [info] Starting imbuement decay for item soulbleeder
[2025-01-07 10:38:02.195] [info] Starting imbuement decay for item soulbleeder
```


**dequip/equip**
```txt
[2025-01-07 10:39:29.798] [info] Stopping imbuement decay for item alicorn headguard
[2025-01-07 10:39:29.798] [info] Stopping imbuement decay for item alicorn headguard
[2025-01-07 10:39:32.812] [info] Starting imbuement decay for item alicorn headguard
[2025-01-07 10:39:32.812] [info] Starting imbuement decay for item alicorn headguard
```

**system is checking imbuement from offline players?**
I'm the only player online and doesn't have `crystal backpack` and `jewelled backpack` in my character
![image](https://github.com/user-attachments/assets/c264538d-813b-4b46-8bf6-c1e9f3e45109)

dudantas committed:
Merge remote-tracking branch 'upstream/main' into dudantas/perf-check-imbuements

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
Refactor imbuement decay logic and add canDecayImbuement

Introduced the canDecayImbuement helper to centralize and clarify the conditions under which imbuements can decay. Updated stopImbuementDecay and checkImbuementDecay to use this function, improving code readability and maintainability by removing duplicated logic.

dudantas committed:
Add imbuements test fixtures and unit/integration tests

Introduces test fixtures and XML/Lua config for imbuements, adds shared test base, and implements both unit and integration tests for imbuement loading and application. Updates CMakeLists to include new test sources. Also fixes missing include in imbuements.hpp.

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
Improve imbuement decay handling for items

Added checks to ensure imbuement decay only starts for items with imbuements and updated logic to start and stop imbuement decay when equipping or de-equipping items. Also refined inventory loading to handle imbuement decay for relevant items on login.

dudantas committed:
Remove redundant canDecayImbuement check in stopImbuementDecay

Eliminated an unnecessary call to canDecayImbuement in the stopImbuementDecay function, streamlining the logic for handling imbuement decay on items.

dudantas committed:
Merge branch 'main' into dudantas/perf-check-imbuements

dudantas committed:
Guard stop imbuement decay with can decay

dudantas committed:
Improve imbuement decay tracking and timing accuracy

Refactors imbuement decay logic to track last update times per item, ensuring more accurate decay timing. Updates variable names for clarity in player inventory loading and ensures proper cleanup of tracking data when items are removed.

dudantas committed:
Fix imbuement login decay tracking

dudantas committed:
Merge branch 'main' into dudantas/perf-check-imbuements

dudantas committed:
Make ItemTypeScope destructor noexcept and handle resize errors

Updated the ItemTypeScope destructor to be noexcept and added error handling for potential exceptions during vector resize in cleanup. This ensures test cleanup does not throw and leaves the vector unchanged if resize fails.

dudantas committed:
Add explicit discard of emplace_back return values

Added (void) casts to emplace_back calls to explicitly discard their return values and silence compiler warnings. Also improved exception safety and comments in test code, and made minor style adjustments for clarity.

dudantas committed:
Merge branch 'main' into dudantas/perf-check-imbuements

dudantas committed:
Merge branch 'main' into dudantas/perf-check-imbuements

github-actions[bot] committed:
style: auto-formatting (clang/stylua/cmake)

dudantas committed:
Remove redundant catch block in imbuements test

Eliminated an unnecessary catch-all exception handler in the imbuements_it.cpp test cleanup code, improving clarity and exception handling.

dudantas committed:
Add Cppcheck cache and update reviewdog config

Introduces a cache step for Cppcheck in the GitHub Actions workflow to improve performance. Updates the .reviewdog.yml configuration to use parallel jobs, limit enabled checks, and specify a build directory for Cppcheck.

dudantas committed:
Update vcpkg setup and MSBuild integration in workflow

Specifies vcpkgDirectory for run-vcpkg and adds ForceImportAfterCppProps and ForceImportAfterCppTargets properties to the MSBuild command for improved vcpkg integration in the reusable Windows build workflow.

dudantas committed:
Improve vcpkg integration in build workflow and project

Updated the GitHub Actions workflow to use PowerShell for building and to dynamically construct vcpkg paths, improving maintainability. Also added conditional import of vcpkg.targets in canary.vcxproj to ensure proper integration with vcpkg when available.

github-actions[bot] committed:
style: auto-formatting (clang/stylua/cmake)

dudantas committed:
Remove .cppcheck_cache directory and contents

Deleted all files under the .cppcheck_cache directory to clean up cached static analysis data. This helps reduce repository clutter and ensures that outdated cache files do not interfere with future static analysis runs.

dudantas committed:
Remove trailing whitespace in workflow YAML

Cleaned up unnecessary trailing whitespace in the reusable-build-windows-solution.yml GitHub Actions workflow for improved code style consistency.

dudantas committed:
Update CI workflows and Windows project toolset

Renamed build jobs in CI workflows for clarity and consistency. Changed VCPKG_BINARY_SOURCES to use 'gha' instead of 'x-gha'. Updated Windows build job names and set canary.vcxproj PlatformToolset to v143. Removed vcpkg.targets import from canary.vcxproj.

dudantas committed:
Update CI workflows and set GCC 13 for Linux builds

Removes the global VCPKG_BINARY_SOURCES environment variable from all reusable build workflows. For Linux builds, sets GCC 13 as the compiler and ensures required packages are installed, including a fix for the ninja-build symlink. These changes improve build environment consistency and update compiler versions.

dudantas committed:
Update CI workflows for VCPKG cache and Windows 2025

Set VCPKG_BINARY_SOURCES for improved caching in all build workflows. Switch Windows runners to windows-2025 and enable latest Visual Studio prerelease for solution builds.

dudantas committed:
Add vcpkg.props import to canary.vcxproj

Added an import statement for vcpkg.props in canary.vcxproj to ensure vcpkg integration during builds. Removed unused vcpkg props/targets path variables from the reusable GitHub Actions workflow.

dudantas committed:
Remove VCPKG_BINARY_SOURCES from build workflows

Eliminated the VCPKG_BINARY_SOURCES environment variable from Linux and Windows reusable build workflows to simplify configuration and rely on default vcpkg binary source behavior. Also updated the Visual Studio version in the Windows solution workflow to '18'.

dudantas committed:
Add VS 2026 Build Tools installation to workflow

The workflow now installs Visual Studio 2026 Build Tools (v145 toolset) using Chocolatey before setting up MSBuild. Also updates the MSBuild setup to use the latest Visual Studio version.

dudantas committed:
Update VS 2026 build tools installation step

Replaces the installation of 'visualstudio2026buildtools' with 'visualstudio2026-workload-vctools' in the workflow. Adds '--ignore-package-exit-codes=3010' to handle non-critical exit codes and updates comments for clarity.

dudantas committed:
Consolidate Windows build workflows in CI

Removed the separate reusable-build-windows-solution.yml workflow and merged its logic into reusable-build-windows.yml using a matrix strategy for both CMake and Solution builds. Updated ci.yml to use the unified workflow, simplifying maintenance and reducing duplication.

dudantas committed:
Merge branch 'main' into dudantas/perf-check-imbuements

dudantas committed:
Improve CI workflows and Windows build artifacts

Make several CI and build improvements: set MSBuild OutDir to an artifacts folder and upload that folder (Windows reusable build), remove local Cppcheck caching and adjust auto-format 'add' paths in checks workflow, and tighten cppcheck invocation/exclusions in reviewdog config. Harden vcpkg baseline resolution by peeling tags and failing fast if SHA can't be determined, with clearer logging. Update README to use consolidated CI badge. Fix vcxproj vcpkg import paths (remove props import and escape target path).

dudantas committed:
Merge branch 'main' into dudantas/perf-check-imbuements

dudantas committed:
Run imbuement decay check every 1s

Reduce ImbuementDecay scheduler interval from 60000ms to 1000ms so decay checks run every second. Update both the initial schedule and the reschedule call and adjust the trace log message in src/creatures/players/imbuements/imbuements.cpp.

dudantas committed:
Use weak_ptr map for imbuement decay tracking

Refactor ImbuementDecay to track items by raw Item* key into an unordered_map holding a TrackedImbuementItem (weak_ptr<Item> + lastUpdate) instead of storing shared_ptrs in a set and a separate last-update map. This prevents keeping items alive unnecessarily, handles expired items correctly, and allows safe erase-while-iterating logic. Removed m_itemLastUpdate and m_lastUpdateTime in favor of per-entry lastUpdate and simplified start/stop/check logic (use emplace/erase, update lastUpdate on successful checks). Overall fixes lifetime bugs and simplifies decay bookkeeping.

github-actions[bot] committed:
style: auto-formatting (clang/stylua/cmake)

dudantas committed:
Merge branch 'main' into dudantas/perf-check-imbuements

dudantas committed:
Merge branch 'main' into dudantas/perf-check-imbuements
